politeiawww/user/mysql/mysql.go

```diff
@@ -0,0 +1,1028 @@
+		if err != nil {
+			return fmt.Errorf("upate: %v", err)
+		}
+	} else {
+		_, err := m.userDB.ExecContext(ctx,
+			`INSERT INTO sessions
+		(k, userID, createdAt, sBlob)
+		VALUES ($1, $2, $3, $4)`,
+			session.Key, session.UserID, session.CreatedAt, session.Blob)
+		if err != nil {
+			return fmt.Errorf("create: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// Get a session by its ID. Returns a user.ErrorSessionNotFound if the given
+// session ID does not exist
+//
+// SessionGetByID satisfies the Database interface.
+func (m *mysql) SessionGetByID(sid string) (*user.Session, error) {
+	log.Tracef("SessionGetByID: %v", sid)
+
+	if m.isShutdown() {
+		return nil, user.ErrShutdown
+	}
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	var blob []byte
+	err := m.userDB.QueryRowContext(ctx, "SELECT sBlob FROM sessions WHERE k = $1",
+		hex.EncodeToString(util.Digest([]byte(sid)))).
+		Scan(&blob)
+	switch {
+	case err == sql.ErrNoRows:
+		return nil, user.ErrUserNotFound
+	case err != nil:
+		return nil, err
+	}
+
+	b, _, err := m.decrypt(blob)
+	if err != nil {
+		return nil, err
+	}
+	return user.DecodeSession(b)
+}
+
+// SessionDeleteByID deletes the session with the given id.
+//
+// SessionDeleteByID satisfies the Database interface.
+func (m *mysql) SessionDeleteByID(sid string) error {
+	log.Tracef("SessionDeleteByID: %v", sid)
+
+	if m.isShutdown() {
+		return user.ErrShutdown
+	}
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	_, err := m.userDB.ExecContext(ctx, "DELETE FROM sessions WHERE k = $1",
+		hex.EncodeToString(util.Digest([]byte(sid))))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SessionsDeleteByUserID deletes all sessions for the given user ID, except
+// the session IDs in exemptSessionIDs.
+//
+// SessionsDeleteByUserID satisfies the Database interface.
+func (m *mysql) SessionsDeleteByUserID(uid uuid.UUID, exemptSessionIDs []string) error {
+	log.Tracef("SessionsDeleteByUserID: %v %v", uid.String(), exemptSessionIDs)
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	// Session primary key is a SHA256 hash of the session ID.
+	exempt := make([]string, 0, len(exemptSessionIDs))
+	for _, v := range exemptSessionIDs {
+		exempt = append(exempt, hex.EncodeToString(util.Digest([]byte(v))))
+	}
+
+	// Using an empty NOT IN() set will result in no records being
+	// deleted.
+	if len(exempt) == 0 {
+		_, err := m.userDB.
+			ExecContext(ctx, "DELETE FROM sessions WHERE userID = $1", uid.String())
+		return err
+	}
+
+	_, err := m.userDB.
+		ExecContext(ctx, "DELETE FROM sessions WHERE usedID = ? AND key NOT IN (?)",
+			uid.String(), exempt)
+	return err
+}
+
+// RegisterPlugin registers a plugin.
+func (m *mysql) RegisterPlugin(p user.Plugin) error {
+	log.Tracef("RegisterPlugin: %v %v", p.ID, p.Version)
+
+	if m.isShutdown() {
+		return user.ErrShutdown
+	}
+
+	// Setup plugin tables
+	var err error
+	switch p.ID {
+	case user.CMSPluginID:
+		// XXX add the following:
```

Comment:
I would take this commented code out for now.  ayp can add it back in if he needs it.